lib/mp4/transmuxer.js

```diff
@@ -274,24 +340,318 @@ VideoSegmentStream = function(track) {
     // throwing away hundreds of media segment fragments
     boxes = new Uint8Array(moof.byteLength + mdat.byteLength);
 
-    // bump the sequence number for next time
+    // Bump the sequence number for next time
     sequenceNumber++;
 
     boxes.set(moof);
     boxes.set(mdat, moof.byteLength);
 
-    clearDtsInfo(track);
     this.trigger('data', {track: track, boxes: boxes});
 
+    this.resetStreamStage_();
+
+    // Continue with the flush process now
+    this.trigger('done');
+  };
+
+  this.resetStreamStage_ = function() {
+    clearDtsInfo(track);
+
     // reset config and pps because they may differ across segments
     // for instance, when we are rendition switching
     config = undefined;
     pps = undefined;
+  };
 
-    // Continue with the flush process now
-    this.trigger('done');
+  // Search for a candidate Gop for gop-fusion from the gop cache and
+  // return it or return null if no good candidate was found
+  this.getGopForFusion_ = function (nalUnit) {
+    var
+      halfSecond = 45000, // Half-a-second in a 90khz clock
+      allowableOverlap = 9009, // About 3 frames
+      nearestDistance = Infinity,
+      dtsDistance,
+      nearestGopObj,
+      currentGop,
+      currentGopObj,
+      i;
+
+    // Search for the GOP nearest to the beginning of this nal unit
+    for (i = 0; i < this.gopCache.length; i++) {
+      currentGopObj = this.gopCache[i];
+      currentGop = currentGopObj.gop;
+
+      // Reject Gops with different SPS or PPS
+      if (!(track.pps && compareArrays(track.pps[0], currentGopObj.pps[0])) ||
+          !(track.sps && compareArrays(track.sps[0], currentGopObj.sps[0]))) {
+        continue;
+      }
+
+      // Reject Gops that would require a negative baseMediaDecodeTime
+      if (currentGop.dts < track.timelineStartInfo.dts) {
+        continue;
+      }
+
+      // The distance between the end of the gop and the start of the nalUnit
+      dtsDistance = (nalUnit.dts - currentGop.dts) - currentGop.duration;
+
+      // Only consider GOPS that start before the nal unit and end within
+      // a half-second of the nal unit
+      if (dtsDistance >= -allowableOverlap &&
+          dtsDistance <= halfSecond) {
+
+        // Always use the closest GOP we found if there is more than
+        // one candidate
+        if (!nearestGopObj ||
+            nearestDistance > dtsDistance) {
+          nearestGopObj = currentGopObj;
+          nearestDistance = dtsDistance;
+        }
+      }
+    }
+
+    if (nearestGopObj) {
+      return nearestGopObj.gop;
+    }
+    return null;
+  };
+
+  // create the default sample
+  // see ISO/IEC 14496-12:2012, section 8.6.4.3
+  this.createDefaultSample_ = function() {
+    return {
+      size: 0,
+      flags: {
+        isLeading: 0,
+        dependsOn: 1,
+        isDependedOn: 0,
+        hasRedundancy: 0,
+        degradationPriority: 0
+      }
+    };
+  };
+
+  this.filterLeadingNonIFrames_ = function(gops) {
```

Comment:
I think this function name is a bit misleading now that it operates on gops. What about something like extendLeadingGop()?